src/primitives/decode/segwit.rs

```diff
@@ -0,0 +1,112 @@
+// SPDX-License-Identifier: MIT
+
+//! Decoding of Segregated Witness address encoded as bech32 strings as specified by [BIP-173] (and [BIP-350]).
+
+use core::{fmt, str};
+
+use crate::primitives::checksum::Checksum;
+use crate::primitives::decode::{ByteIter, Fe32Iter};
+use crate::primitives::gf32::Fe32;
+use crate::primitives::hrp::Hrp;
+use crate::write_err;
+
+/// An HRP string that has been parsed and had the checksum validated.
+///
+/// Pre-parsing an HRP string only checks validity of the characters, it does not validate the
+/// checksum in any way - to validate convert to [`Parsed`].
+#[derive(Debug)]
+pub struct Parsed<'s> {
+    /// The parsed HRP and data.
+    inner: super::Parsed<'s>,
+    /// The first byte of `self.inner.data`.
+    witness_version: u8,
+}
+
+impl<'s> Parsed<'s> {
+    /// Parses an HRP string, treating the first data character as a witness version.
+    ///
+    /// This version byte does not appear in the extracted binary data, but is covered
+    /// by the checksum. It can be accessed with [`Self::witness_version`] and is also
+    /// returned from this constructor as a convenience.
+    pub fn new<Ck: Checksum>(s: &'s str) -> Result<Self, Error> {
+        let validated = super::Parsed::new::<Ck>(s)?;
+
+        // Unwrap ok since check_characters (in `Self::new`) checked the bech32-ness of this char.
+        let witness_version = Fe32::from_char(validated.data[0].into()).unwrap().to_u8();
+
+        // From BIP-173:
+        // > Re-arrange those bits into groups of 8 bits. Any incomplete group at the
+        // > end MUST be 4 bits or less, MUST be all zeroes, and is discarded.
+        let data_len = validated.data.len() - 1; // -1 for witness version.
+        if data_len * 5 % 8 > 4 {
+            return Err(Error::InvalidDataLength);
+        }
+
+        Ok(Self { inner: validated, witness_version })
+    }
+
+    /// Returns the witness version if present.
+    pub fn witness_version(&self) -> u8 { self.witness_version }
+
+    /// Returns the human-readable part.
+    pub fn hrp(&self) -> Hrp { self.inner.hrp() }
+
+    /// Returns an iterator over the byte data encoded by the HRP string (excluding the HRP, witness
+    /// version byte if any, and checksum).
+    pub fn byte_iter(&self) -> ByteIter { self.inner.byte_iter() }
+
+    /// Returns an iterate over field elements of the data encoded by the HRP string (excluding the
+    /// HRP, witness version byte if any, and checksum).
+    pub fn fe32_iter(&self) -> Fe32Iter { self.inner.fe32_iter() }
```

Comment:
In 5c9b549a3738cec82f633c14e274854b82edce45:

I think this should be dropped. It's a little confusing to have an iterator over field elements that doesn't include the HRP data or checksum, and it's not clear to me what the value is. Note that this is not used anywhere.

Reply:
Same with the other one, on the othe `Parsed`.

Reply:
Dropped!